dlib/internal/integer_utils.cpp

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2019 Paul Dreik (github@pauldreik.se)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#include "integer_utils.h"
+
+#include <stdexcept>
+#include <cstdint>
+#include <limits>
+
+namespace dlib
+{
+namespace detail {
+long safe_multiply(long a, long b) {
+    long ret;
+    bool overflow=false;
+#ifdef DLIB_HAS_BUILTIN_MUL_OVERFLOW
+    // works for clang and gcc
+    overflow=__builtin_mul_overflow(a,b,&ret);
+#else
+    using small=long;
+    using large=std::intmax_t;
+    if(sizeof(small)<sizeof(large)) {
+        // fallback: promoting to a larger type
+        // This works for MSVC (yet only tested on compiler explorer).
+        auto tmp=large{a}*b;
+        if(tmp<=std::numeric_limits<small>::max() && tmp>=std::numeric_limits<small>::min()) {
+            overflow=false;
+            ret=tmp;
+        } else {
+            overflow=true;
+        }
+    } else {
+       // unsupported compiler
+       throw std::logic_error("signed integer overflow check not implemented on this platform");
```

Comment:
This code must be made to work everywhere. 

I think you should be able to do all this much more simply, by essentially doing:
```
result = a*b;
if (result/b != a) then there was overflow
```
You have to do it in unsigned arithmetic since signed overflow is undefined behavior in C++.  But you can accomplish this here by converting the signed integers into unsigned, doing the multiply, and converting it back to signed, multiplying by -1 if needed, and checking that the conversion to signed is equal to the unsigned value (accounting for the -1).